### DISCUSSION TITLE
Why there is an implementation of IntoResponse for `()`

### DISCUSSION THREAD

**Discussion by miaomiaowu0428** [tokio-rs/axum#2845] [Q&A] [answered]
### Summary

Just as I mentioned in the title, I found this in axum:
```rust
impl IntoResponse for () {
    fn into_response(self) -> Response {
        Body::empty().into_response()
    }
}
```

What if someone simply want's to return `"success"` like this
```rust
 fn some_func() -> impl IntoResponse {
    "success".into_response()
}
```
but make a mistake like this:
```rust
 fn some_func() -> impl IntoResponse {
    "success".into_response();
}
```

then he/she will found this function always response a empty body with code 200
the mistake is easy to happen too inconspicuous to be spotted at the first time

If someone want's to make a response with an empty body, he/she shoule use `StatusCode.into_response()`, right?

### axum version

0.7.5

**Comment by mladedav** [comment]
I guess this is a quality of life for simple handlers like `async fn handle() { println!("I'm handling things!"); }`

Some tests use it for example like this: `let two = Router::new().route("/baz", get(|| async {}));`

But I have to agree this might be something worth at least considering deprecating. In the meanwhile I'd suggest using a concrete return type rather than RPIT. This also shouldn't be an issue if you return from multiple places an make a mistake just once like
```rust
 fn some_func() -> impl IntoResponse { // ERROR: mismatched types expected struct `http::Response<body::Body>` found unit type `()`
    if false {
        return "failure".into_response();
    }
    "success".into_response();
}
```

As a side note, since these types already implement `IntoResponse`, there's no need to call `into_response` on them explicitly. Just return `"success"`, `()`, `StatusCode::OK` or whatherever else. You should have to call this explicitly only if you want to return different types from different places, effectively returning resolved `Response`s instead.

**Reply by jplatte** [reply]
Maybe we should add `#[must_use]` to `fn into_response`.

**Comment by jplatte** [comment, marked as answer]
Once https://github.com/tokio-rs/axum/pull/2846 is merged and released, your example and similar code will generate a warning:

![Screenshot_2024-07-24_225933](https://github.com/user-attachments/assets/b1fe97ac-aabd-4baa-99e6-d701260146f0)

**Reply by miaomiaowu0428** [reply]
nice!
